File: app/src/main/java/fieldmarshal/repobrowse/util/LinkPager.java
package fieldmarshal.repobrowse.util;

import android.support.annotation.NonNull;
import okhttp3.Headers;

/**
 * Created by fieldmarshal on 15.11.17.
 */

public class LinkPager {
    private static final String DELIM_LINKS = ",";

    private static final String DELIM_LINK_PARAM = ";";
    private static final String HEADER_LINK = "Link";
    private static final String META_REL = "rel";
    private static final String META_FIRST = "first";
    private static final String META_LAST = "last";
    private static final String META_NEXT = "next";
    private static final String META_PREV = "prev";
    private static final String HEADER_LAST = "X-Last";
    private static final String HEADER_NEXT = "X-Next";

    private String first;
    private String last;
    private String next;
    private String prev;


    /**
     * Parse links from executed method
     */
    public LinkPager(Headers headers) {
        String linkHeader = headers.get(HEADER_LINK);
        if (linkHeader != null) {
            String[] links = linkHeader.split(DELIM_LINKS);
            for (String link : links) {
                String[] segments = link.split(DELIM_LINK_PARAM);
                if (segments.length < 2)
                    continue;

                String linkPart = segments[0].trim();
                if (!linkPart.startsWith("<") || !linkPart.endsWith(">"))
                    continue;
                linkPart = linkPart.substring(1, linkPart.length() - 1);

                for (int i = 1; i < segments.length; i++) {
                    String[] rel = segments[i].trim().split("=");
                    if (rel.length < 2 || !META_REL.equals(rel[0]))
                        continue;

                    String relValue = rel[1];
                    if (relValue.startsWith("\"") && relValue.endsWith("\""))
                        relValue = relValue.substring(1, relValue.length() - 1);

                    if (META_FIRST.equals(relValue))
                        first = linkPart;
                    else if (META_LAST.equals(relValue))
                        last = linkPart;
                    else if (META_NEXT.equals(relValue))
                        next = linkPart;
                    else if (META_PREV.equals(relValue))
                        prev = linkPart;
                }
            }
        } else {
            next = headers.get(HEADER_NEXT);
            last = headers.get(HEADER_LAST);
        }
    }

    // Better when links are non-null
    public @NonNull String getFirst() {
        return first;
    }

    public @NonNull String getLast() {
        return last;
    }

    public @NonNull String getNext() {
        return next;
    }

    public Integer getPage(String url) {
        final String[] trimmed = url.split("page=");
        return Integer.getInteger(trimmed[1]);
    }

    public String getPrev() {
        return prev;
    }
}

File: app/src/main/java/fieldmarshal/repobrowse/util/LinkPager.kt
package fieldmarshal.repobrowse.util

import okhttp3.Headers

/**
 * Created by fieldmarshal on 15.11.17.
 */

class LinkPager
/**
 * Parses links for the executed method
 * @param headers headers to be parsed
 */
(headers: Headers) {

    var first: String? = null
        private set
    var last: String? = null
        private set
    var prev: String? = null
        private set
    var next: String? = null
        private set

    private var nextPage = 2

    init {
        val linkHeader = headers.get(HEADER_LINK)
        if (linkHeader != null) {
            val links = linkHeader.split(DELIM_LINKS.toRegex())
                    .dropLastWhile { it.isEmpty() }.toTypedArray()
            for (link in links) {
                val segments = link.split(DELIM_LINK_PARAM.toRegex())
                        .dropLastWhile { it.isEmpty() }.toTypedArray()
                if (segments.size < 2)
                    continue

                var linkPart = segments[0].trim { it <= ' ' }
                if (!linkPart.startsWith("<") || !linkPart.endsWith(">"))
                    continue
                linkPart = linkPart.substring(1, linkPart.length - 1)

                for (i in 1 until segments.size) {
                    val rel = segments[i].trim { it <= ' ' }.split("=".toRegex())
                            .dropLastWhile { it.isEmpty() }.toTypedArray()
                    if (rel.size < 2 || META_REL != rel[0])
                        continue

                    var relValue = rel[1]
                    if (relValue.startsWith("\"") && relValue.endsWith("\""))
                        relValue = relValue.substring(1, relValue.length - 1)

                    if (META_FIRST == relValue)
                        first = linkPart
                    else if (META_LAST == relValue)
                        last = linkPart
                    else if (META_NEXT == relValue)
                        next = linkPart
                    else if (META_PREV == relValue)
                        prev = linkPart
                }
            }
        } else {
            next = headers.get(HEADER_NEXT)
            last = headers.get(HEADER_LAST)
        }
    }

    fun getNextPage(): Int {
        val trimmed = next!!.split("page=".toRegex())
                .dropLastWhile { it.isEmpty() }.toTypedArray()
        nextPage = Integer.getInteger(trimmed[1])!!
        return nextPage
    }

    companion object {
        private val DELIM_LINKS = ","
        private val DELIM_LINK_PARAM = ";"

        private val HEADER_LINK = "Link"
        private val HEADER_LAST = "X-Last"
        private val HEADER_NEXT = "X-Next"
        private val META_REL = "rel"
        private val META_FIRST = "first"
        private val META_LAST = "last"
        private val META_PREV = "prev"
        private val META_NEXT = "next"
    }
}
